File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Ros/Service/CdrHelpers.swift
//
//  CdrHelpers.swift
//  RaspberryPiClient
//
//  CDR 形式の共通ヘルパー（複数サービスで共用）
//  Ref: https://www.omg.org/spec/DDS-XTypes/
//

import Foundation

/// CDR 形式の共通ヘルパー（複数サービスで共用）
enum CdrHelpers {
    /// CDR encapsulation header (CDR_BE, little-endian)
    static let encapsulationHeader: [UInt8] = [0x00, 0x01, 0x00, 0x00]

    /// CDR string をエンコード（4byte length + UTF-8 + NUL）
    /// - Parameter value: エンコードする文字列（nil の場合は空文字として扱う）
    /// - Returns: CDR 形式の string データ
    static func encodeCdrString(_ value: String?) -> Data {
        let nameBytes = (value ?? "").data(using: .utf8) ?? Data()
        let length = UInt32(nameBytes.count + 1)
        var data = Data()
        data.append(contentsOf: withUnsafeBytes(of: length.littleEndian) { Array($0) })
        data.append(nameBytes)
        data.append(0)
        return data
    }

    /// CDR string をデコード
    /// - Parameters:
    ///   - data: CDR データ
    ///   - offset: string の開始位置（4byte length の先頭）
    /// - Returns: デコードされた文字列。範囲外または不正なデータの場合は nil
    static func decodeCdrString(from data: Data, at offset: Int) -> String? {
        guard data.count >= offset + 4 else { return nil }
        let lengthBytes = data.subdata(in: offset ..< (offset + 4))
        let length = lengthBytes.withUnsafeBytes { $0.load(as: UInt32.self).littleEndian }
        guard data.count >= offset + 4 + Int(length), length > 0 else { return nil }
        let stringData = data.subdata(in: (offset + 4) ..< (offset + 4 + Int(length) - 1))
        return String(data: stringData, encoding: .utf8)
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/ros/cdr/CdrHelpers.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.ros.cdr

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Ref: [DDS-XTypes](https://www.omg.org/spec/DDS-XTypes/)
 */
object CdrHelpers {

    /**
     * CDR カプセル化ヘッダー（4 バイト）。
     * バイト 0–1: 0x00 0x01 → CDR_LE。デコード時はこの 4 バイトをスキップしてペイロードを読む。
     */
    val encapsulationHeader: ByteArray = byteArrayOf(0x00, 0x01, 0x00, 0x00)

    /**
     * CDR string をデコードする
     *
     * @param data CDR バイナリ
     * @param offset **4 バイトの uint32 LE 長さフィールドの先頭**（string ブロックの開始位置）
     * @return UTF-8 文字列（末尾 NUL は含めない）。範囲外または不正なデータのときは null
     */
    fun decodeCdrString(data: ByteArray, offset: Int): String? {
        if (data.size < offset + 4) return null
        val length = readUInt32LittleEndian(data, offset)
        if (length == 0u) return null
        if (data.size < offset + 4 + length.toInt()) return null
        val start = offset + 4
        val end = offset + 4 + length.toInt() - 1
        val stringBytes = data.copyOfRange(start, end)
        return stringBytes.decodeToString()
    }

    /**
     * CDR string をエンコードする。
     *
     * uint32 LE 長さフィールド（null 終端を含む）＋ UTF-8 バイト列 ＋ null バイト。
     *
     * @param value エンコードする文字列
     * @return CDR string バイト列
     */
    fun encodeCdrString(value: String): ByteArray {
        val strBytes = value.encodeToByteArray()
        val length = strBytes.size + 1  // null 終端を含む長さ
        val buf = ByteBuffer.allocate(4 + length).order(ByteOrder.LITTLE_ENDIAN)
        buf.putInt(length)
        buf.put(strBytes)
        buf.put(0x00.toByte())
        return buf.array()
    }

    private fun readUInt32LittleEndian(data: ByteArray, offset: Int): UInt {
        return ByteBuffer.wrap(data, offset, 4)
            .order(ByteOrder.LITTLE_ENDIAN)
            .int
            .toUInt()
    }
}
